### PR TITLE
[WIP] Feat/interactive root picker

### PR DESCRIPTION
**Pull request by mjlbach**
Closes https://github.com/neovim/nvim-lspconfig/issues/1028

### CODE REVIEW

File: lua/lspconfig/ui/prompt.lua [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	["p"]=function() return parent_dir end,
+	["c"]=M.directory_picker,
+	["i"]=function() return nil end,
+    }
+
+    local user_input
+    local user_choice
+    local root_directory
+
+    while true do
```
Comment (mjlbach) [inline lua/lspconfig/ui/prompt.lua:36]: @folke Do you have a better solution for blocking input? I need to return a result from this function so I can't just map the callbacks and be done with it.

Comment (folke) [reply lua/lspconfig/ui/prompt.lua]: That's pretty much what I do in WhichKey too.

The only thing I would change is:

```lua
  local ok, n = pcall(vim.fn.getchar)
 if not ok then break end
```

That properly handles keyboard interrupts too (user pressing `<control>c`)

**Comment by folke** [comment]
Let me know once it makes sense to test and do a code review. Looking good!

**Comment by mjlbach** [comment]
> CONFIG.md is auto-generated: edit the lua source file instead (and re-read the PR template).

you have made an enemy today, my child.

File: lua/lspconfig/efm.lua [hunk 2]
```diff
@@ -7,9 +7,7 @@ local bin_name = "efm-langserver"
 configs[server_name] = {
   default_config = {
     cmd = { bin_name },
-    root_dir = function(fname)
-      return util.root_pattern ".git"(fname) or util.path.dirname(fname)
-    end,
+    root_dir = util.root_pattern ".git",
```
Comment (WhyNotHugo) [inline lua/lspconfig/efm.lua:10]: This change breaks for me. Using the current directory if outside a git repository makes EFM work for things that are standalone scripts and alike.

File: lua/lspconfig/gopls.lua [hunk 3]
```diff
@@ -15,9 +15,7 @@ https://github.com/golang/tools/tree/master/gopls
 
 Google's lsp server for golang.
 ]],
-    default_config = {
-      root_dir = [[root_pattern("go.mod", ".git")]],
```
Comment (WhyNotHugo) [inline lua/lspconfig/gopls.lua:19]: The previous version here seems pretty solid, given that `go.mod` exists in the root of all modern `golang` projects.

File: lua/lspconfig/jedi_language_server.lua [hunk 4]
```diff
@@ -1,22 +1,26 @@
...
+}
 
 configs.jedi_language_server = {
   default_config = {
     cmd = { "jedi-language-server" },
     filetypes = { "python" },
-    root_dir = function(fname)
-      return vim.fn.getcwd()
-    end,
+    root_dir = util.root_pattern(unpack(root_files)),
```
Comment (WhyNotHugo) [inline lua/lspconfig/jedi_language_server.lua:16]: Nice! :+1:

File: lua/lspconfig/jsonls.lua [hunk 5]
```diff
@@ -11,7 +11,7 @@ configs[server_name] = {
     init_options = {
       provideFormatter = true,
     },
-    root_dir = util.root_pattern(".git", vim.fn.getcwd()),
+    root_dir = util.find_git_ancestor,
```
Comment (WhyNotHugo) [inline lua/lspconfig/jsonls.lua:14]: Having `cwd` as a fallback is great here. The `JSON` LSP doesn't even have to deal with things like imports or alike, it _just_ checks the current file.

File: lua/lspconfig/solargraph.lua [hunk 6]
```diff
@@ -30,9 +30,7 @@ You can install solargraph via gem install.
 gem install --user-install solargraph
 ```
     ]],
-    default_config = {
-      root_dir = [[root_pattern("Gemfile", ".git")]],
```
Comment (WhyNotHugo) [inline lua/lspconfig/solargraph.lua:34]: `Gemfile` as a fallback here made more sense that not having it.

**Review by WhyNotHugo — COMMENTED** [review]
I'm not a fan of having the interactive picker become the default. It's very invasive: when opening a file, I generally want to start reading / editing right away, not configuring an LSP or dealing with plugins prompting for data.

Configurating neovim and using neovim are two very distinct activities for me, and I don't want to deal with under-the-hood nuances when working on something else.

It should definitely be configurable option, and, IMHO, disabled by default.

Some of the auto-detection fallbacks that are being dropped here were pretty solid, and, personally, I think we'd be dropping some very useful code by deleting them. Is there any rationale for not keeping them, especially the ones that work fine?

File: lua/lspconfig/gopls.lua (hunk 3 above)
Comment (mjlbach) [reply lua/lspconfig/gopls.lua:19]: This is documentation only, it's autogenerated now and does not affect current behavior in any way.

File: lua/lspconfig/jsonls.lua (hunk 5 above)
Comment (mjlbach) [reply lua/lspconfig/jsonls.lua:14]: get_cwd() is completely incorrect, it should be, at best, util.path.dirname

File: lua/lspconfig/solargraph.lua (hunk 6 above)
Comment (mjlbach) [reply lua/lspconfig/solargraph.lua:34]: This is documentation only, it's autogenerated now and does not affect current behavior in any way.

**Comment by mjlbach** [comment]
For those leaving drive-by comments, removing `default_config` from the *documentation* section does not change current behavior. It uses the auto-generated documentation pipeline I pulled to extract the function and paste the text. Many of our currently documented root directories do *not* match the actual code.

**Pull request closed without merging**